File: apps/api/app/routes/tts.py
"""
POST /api/tts
Proxy to Sarvam AI text-to-speech — keeps the API key server-side.
"""
import os
import logging
import httpx
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel

logger = logging.getLogger(__name__)
router = APIRouter()

SARVAM_URL = "https://api.sarvam.ai/text-to-speech"
SARVAM_KEY = os.getenv("SARVAM_API_KEY", "")


class TTSRequest(BaseModel):
    text: str
    speaker: str = "vidya"
    pace: float = 0.85
    pitch: float = 0.0
    loudness: float = 2.2
    language_code: str = "hi-IN"


@router.post("/api/tts")
async def tts_proxy(req: TTSRequest):
    if not SARVAM_KEY:
        raise HTTPException(status_code=503, detail="TTS service not configured")

    if not req.text.strip():
        raise HTTPException(status_code=400, detail="text is required")

    payload = {
        "inputs": [req.text],
        "target_language_code": req.language_code,
        "speaker": req.speaker,
        "pitch": req.pitch,
        "pace": req.pace,
        "loudness": req.loudness,
        "speech_sample_rate": 22050,
        "enable_preprocessing": True,
        "model": "bulbul:v2",
    }

    async with httpx.AsyncClient(timeout=15) as client:
        try:
            res = await client.post(
                SARVAM_URL,
                json=payload,
                headers={
                    "Content-Type": "application/json",
                    "api-subscription-key": SARVAM_KEY,
                },
            )
        except httpx.RequestError as e:
            logger.error("Sarvam TTS request failed: %s", e)
            raise HTTPException(status_code=502, detail="TTS upstream unreachable")

    if not res.is_success:
        logger.error("Sarvam TTS error %s: %s", res.status_code, res.text)
        raise HTTPException(status_code=502, detail=f"TTS upstream error {res.status_code}")

    data = res.json()
    audio_b64 = data.get("audios", [None])[0] or data.get("audio")
    if not audio_b64:
        raise HTTPException(status_code=502, detail="No audio in TTS response")

    return {"audio_b64": audio_b64}

File: apps/web/src/lib/tts.ts
/**
 * TTS — Sarvam AI (Hindi) + Web Speech API (English/fallback).
 * Sarvam "meera" is the warmest, most relaxed Hindi female voice.
 */

const SARVAM_KEY = 'sk_gb0t7hf0_XLAp44zXlvnsfgPwNlRl3X91'
const SARVAM_URL = 'https://api.sarvam.ai/text-to-speech'

let currentAudio: HTMLAudioElement | null = null
const audioCache = new Map<string, string>()

export function stopSpeech() {
  if (currentAudio) {
    currentAudio.pause()
    currentAudio.currentTime = 0
    currentAudio = null
  }
  if ('speechSynthesis' in window) window.speechSynthesis.cancel()
}

export async function prefetchSpeech(text: string, lang?: string) {
  if (!text) return
  const cleanText = text.replace(/\*\*/g, '')
  const language = lang ?? localStorage.getItem('goldeye_lang') ?? 'en'
  if (language === 'hi' && !audioCache.has(cleanText)) {
    try {
      const res = await fetch(SARVAM_URL, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json', 'api-subscription-key': SARVAM_KEY },
        body: JSON.stringify({
          inputs: [cleanText], target_language_code: 'hi-IN', speaker: 'vidya', pitch: 0, pace: 0.85, loudness: 1.5, speech_sample_rate: 22050, enable_preprocessing: true, model: 'bulbul:v2'
        })
      })
      if (res.ok) {
        const data = await res.json()
        if (data.audios?.[0]) audioCache.set(cleanText, data.audios[0])
      }
    } catch (e) {
      console.warn('Failed to prefetch TTS:', e)
    }
  }
}

async function sarvamSpeak(text: string): Promise<void> {
  let b64 = audioCache.get(text)

  if (!b64) {
    const res = await fetch(SARVAM_URL, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'api-subscription-key': SARVAM_KEY,
      },
      body: JSON.stringify({
        inputs: [text],
        target_language_code: 'hi-IN',
        speaker: 'vidya',
        pitch: 0,
        pace: 0.85,
        loudness: 1.5,
        speech_sample_rate: 22050,
        enable_preprocessing: true,
        model: 'bulbul:v2',
      }),
    })

    if (!res.ok) throw new Error(`Sarvam TTS ${res.status}`)

    const data = await res.json()
    b64 = data.audios?.[0]
    if (!b64) throw new Error('No audio returned')
    
    audioCache.set(text, b64)
  }

  stopSpeech()

  const audio = new Audio(`data:audio/wav;base64,${b64}`)
  currentAudio = audio
  audio.addEventListener('ended', () => { if (currentAudio === audio) currentAudio = null })
  await audio.play()
}

function webSpeak(text: string, lang: string): void {
  if (!('speechSynthesis' in window)) return
  window.speechSynthesis.cancel()
  const u = new SpeechSynthesisUtterance(text)
  u.lang = lang
  u.rate = 0.95
  window.speechSynthesis.speak(u)
}

export function speak(text: string, lang?: string): void {
  if (!text) return
  // Strip markdown ** bolding before speaking
  const cleanText = text.replace(/\*\*/g, '')
  const language = lang ?? localStorage.getItem('goldeye_lang') ?? 'en'
  if (language === 'hi') {
    sarvamSpeak(cleanText).catch(() => webSpeak(cleanText, 'hi-IN'))
    return
  }
  webSpeak(cleanText, 'en-US')
}
